clang/lib/StaticAnalyzer/Checkers/StreamChecker.cpp

```diff
@@ -763,6 +779,11 @@ void StreamChecker::evalFreadFwrite(const FnDescription *Desc,
     return;
   }
 
+  // At read, invalidate the buffer in any case of error or success,
+  // except if EOF was already present.
+  if (IsFread && (OldSS->ErrorState != ErrorFEof))
+    State = escapeArgs(State, C, Call, {0});
```

Comment:
From what I can tell, the lambda `UpdateBufferRegionForFread` is already used to take care of this in a more fine-grained manner. For instance:

```cpp
int buffer[10];
buffer[5] = 42;
if (1 == fread(buffer, sizeof(int), 5, fd)) {
    assert(buffer[5] == 42);
}
```

Before this change, the assertion would pass, since lambda took `nmemb` into account. With this change, the whole buffer is invalidated.

Reply:
The current solution invalidates the whole buffer (if I think correctly) because the `SVal` of the buffer is passed to `invalidateRegions`. Is there a solution to invalidate a partial buffer (of specific size)?

Reply:
It turns out I was comparing with our downstream copy of v17, where `UpdateBufferRegionForFread` is defined.
So it can be done for the case of arrays. I have double-checked with the original author and he agrees it can be upstreamed. Let me uplift the patch and then I will share it with you.

Reply:
Here it is: https://gist.github.com/alejandro-alvarez-sonarsource/48edec4debc8912a6485f989b2a6f0db

I have rebased it on top of 4f12f47550eee85447c9ec37d27a20c6593d3d40 and run check-clang to make sure I didn't break anything.

Reply:
This could be a future improvement (after this patch). I am not sure if this works always correct, because it looks like that the `ElementCount` value at `fread` is used but not the "size" argument (size of one element). It is assumed that the specified element size is always the same as one array element, but this can be different.

Reply:
Ok, let’s do that, and send this patch as a separate pr. @steakhal LGTM.